woocommerce/woocommerce
# CYS - AI Improvements

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





On this PR, a series of changes are made to improve AI content generation and image management, including:

## JSON Mode Enabled

- Enable [JSON Mode](https://platform.openai.com/docs/guides/text-generation/json-mode) to ensure the AI  model is constrained to only generate strings that parse into valid JSON objects. In practice, this represents an improvement in consistency and performance, as we won't have to rely as much on the built in retry mechanism to ensure the content was generated correctly.
 
## Improvements for images assigned to products

- Reduce the size of the images uploaded to the media library for products. While the `/wpcom/v2/external-media/list/pexels` endpoint provides a series of predefined image sizes for usage, none of them were ideal for our use-case when it comes to products, with the images accessible via `URL` (and `['thumbnails']['large']`) being too big and the images accessible via `['thumbnails']['medium']` and `['thumbnails']['small']` being too  low quality. Initially, we were relying on the highest quality possible, which would work as expected most of the time. Still, on occasion, when the returned image was too big, it could trigger memory exhaustion errors on servers with limited resources, such as JN, which we have been using for testing. This change ensures that images still have a good resolution while at the same time ensuring that they have a reasonable size for downloading and uploading to the media library without any problems.

- Similarly to what was already implemented on `ai/patterns` endpoint, created a transient to store images fetched from Pexels and make them accessible whenever a request to `ai/product` is made but for some reason interrupted due to unavailability of AI or equivalent, avoiding making additional requests to Pexels and AI to generate a search term and fetch images that were already successfully fetched for that execution in particular. This change should improve performance and also reduce costs with AI.

## Updated Prompts

- All prompts were updated for improved consistency, with specific roles being given to AI depending on the expected output.
- The prompt for generating the search term to fetch images from Pexels was updated to return multiple words for the search instead of just one while avoiding generic verbiage that could cause inconsistencies on the selected images.
- The prompt for generating product content was updated to rely on the search term used to fetch the images from Pexels for improved accuracy.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new WooCommerce installation with this version.
2. Ensure Jetpack is installed and your site is connected with JP.
3. Ensure you have the WooCommerce Beta Tester plugin installed and activated (available on this monorepo).
4. Ensure you have the Woo AI plugin installed and activated (available on this monorepo).
5. Head over to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` and enable `customize-store` feature flag:

<img width="1233" alt="Screenshot 2023-10-22 at 10 32 24" src="https://github.com/woocommerce/woocommerce-blocks/assets/15730971/816867c6-34e5-40a1-a87d-f4a7cda46429">

6. Head over to WooCommerce -> Home and click on "Start Customizing":

<img width="1232" alt="Screenshot 2023-12-14 at 10 14 00" src="https://github.com/woocommerce/woocommerce/assets/15730971/5c2258e2-14de-4c16-8dae-3bb0f330a61f">


7. Click Design with AI button
8. Fill out a business description and proceed
9. Make sure everything works as expected, and you can have your store with AI-generated content without problems.


https://github.com/woocommerce/woocommerce/assets/15730971/6028d7e3-e9ad-439b-9894-4300d1ac12ff





### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Enhancement - Improve the AI content generation experience within the CYS onboarding flow.

#### Comment 

</details>
